From aed28351f4ce05b557d90b5257ac7ca651bcb11e Mon Sep 17 00:00:00 2001
From: Mark Harfouche <mark.harfouche@gmail.com>
Date: Fri, 17 Jul 2020 20:06:33 -0400
Subject: [PATCH] Update location of hid headers on windows

---
 teensy_loader_cli.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/teensy_loader_cli.c b/teensy_loader_cli.c
index 6e78f1c..ec0bd30 100644
--- a/teensy_loader_cli.c
+++ b/teensy_loader_cli.c
@@ -370,8 +370,8 @@ int soft_reboot(void)
 // http://msdn.microsoft.com/en-us/library/ms790932.aspx
 #include <windows.h>
 #include <setupapi.h>
-#include <ddk/hidsdi.h>
-#include <ddk/hidclass.h>
+#include <hidsdi.h>
+#include <hidclass.h>
 
 HANDLE open_usb_device(int vid, int pid)
 {
